foundry.md

```diff
@@ -0,0 +1,294 @@
+              ~> #setBuffer(RESULT_ADDR_HANDLE, NEWADDR)
+                 ...
+        </instrs>
+        <commands> (. 
+                => createAccount(NEWADDR)
+                ~> setAccountOwner(NEWADDR, OWNER)
+                ~> setAccountCode(NEWADDR, MODULE)
+                ~> callContract(NEWADDR, "init", mkVmInputDeploy(OWNER, VALUE, ARGS, GAS, 0))
+                ) ...
+        </commands>
+        <account>
+           <address> OWNER </address>
+           <nonce> NONCE </nonce>
+           ...
+        </account>
+        <newAddresses> ... tuple(OWNER, NONCE) |-> NEWADDR:Bytes ... </newAddresses>
+        <wasmStore> ... PATH |-> MODULE </wasmStore>
+
+    rule [instr-deployContract-err]:
+        <instrs> foundryDeployContract(_, _, _, _, _, _)
+              => #throwException(ExecutionFailed, "Could not deploy contract")
+                 ...
+        </instrs>
+      [owise]
+
+```
+
+### Get/set storage
+
+```k
+    rule [hostCall-getStorage]:
+        <instrs> hostCall ( "env" , "getStorage" , [ i32  i32  i32  .ValTypes ] -> [ .ValTypes ] )
+              => foundryGetStorage( getBuffer(OWNER_HANDLE), getBuffer(KEY_HANDLE), DEST_HANDLE)
+                 ...
+        </instrs>
+        <locals>
+          0 |-> <i32> OWNER_HANDLE
+          1 |-> <i32> KEY_HANDLE
+          2 |-> <i32> DEST_HANDLE
+        </locals>
+        <callee> #foundryRunner </callee>
+
+    syntax InternalInstr ::= foundryGetStorage(BytesResult, BytesResult, Int)
+ // -------------------------------------------------------------------------
+    rule [testapi-getStorage]:
+        <instrs> foundryGetStorage(OWNER:Bytes, KEY:Bytes, DEST_HANDLE)
+              => #setBuffer(DEST_HANDLE, VAL)
+                 ...
+        </instrs>
+        <account>
+          <address> OWNER </address>
+          <storage> ...  wrap(KEY) Bytes2Bytes|-> wrap(VAL) ... </storage>
+          ...
+        </account>
+
+    rule [testapi-getStorage-err]:
+        <instrs> foundryGetStorage(_, _, DEST_HANDLE)
+              => #setBuffer(DEST_HANDLE, .Bytes)
+                 ...
+        </instrs>
+      [owise]
+
+```
+
+### Assertions/assumptions
+
+```k
+    rule [hostCall-assertBool]:
+        <instrs> hostCall ( "env" , "assertBool" , [ i32 .ValTypes ] -> [ .ValTypes ] )
+              => #if P =/=Int 0
+                 #then .K
+                 #else #throwException(ExecutionFailed, "assertion failed")
+                 #fi
+                 ...
+        </instrs>
+        <locals>
+          0 |-> <i32> P
+        </locals>
+
+
+    rule [hostCall-assumeBool]:
+        <instrs> hostCall ( "env" , "assumeBool" , [ i32 .ValTypes ] -> [ .ValTypes ] )
+              => #assume(P) ...
+        </instrs>
+        <locals>
+          0 |-> <i32> P
+        </locals>
+
+    syntax IternalInstr ::= #assume(Int)
+ // ------------------------------------
+    rule [assume]:
+        <instrs> #assume(P) => . ... </instrs>
+      ensures P =/=Int 0
```

Comment:
I think that the Foundry code does the same, and, until recently, I agreed with implementing `assume` with an `ensures` clause, but this actually creates an invalid axiom, and the semantics becomes inconsistent.

The main issue is that this axiom is encoded as `LHS => RHS #And P =/=Int 0`, and, when P ==0, this becomes `LHS => RHS #And #Bottom`, i.e. it becomes `LHS => #Bottom`, which is the same as `LHS -> (next) #Bottom`, i.e. `LHS -> #Bottom`, i.e. `LHS` must be `#Bottom`. However, this is not true here, so the entire axiom is `#Bottom`.

See this Slack thread for more details on a similar issue:
https://runtimeverification.slack.com/archives/CC360GUTG/p1691576058510699

We could implement this as a pair of two rules (I'm removing cell names in order to keep it short), or something similar:
```
rule #assume(P) => . requires P =/=Int 0
rule #assume(P) => #endFoundryImmediately requires P ==Int 0
```
where `#endFoundryImmediately` finishes execution instantly with a success status.

I think that the Foundry team was concerned about generating too many splits with the solution above. I'm not sure how much of an issue that actually is, but I don't know of a better solution anyway.